インクリメンタルサーチ

avaklee opened the pull request:
# WHAT

見た目の修正

# WHY

読みやすく・管理しやすくするため

avaklee committed:
インクリメンタルサーチを実装

# WHAT

インクリメンタルサーチ実装 Step 1-7 カリキュラム

#WHY

Userが名前を検索した時に、入れた文字に関連するユーザーを表示する。使いやすくさせるため

avaklee committed:
見た目、簡略化の修正

# WHAT

見た目の修正

#WHY

読みやすく・管理しやすくするため

avaklee committed:
Scopeなどの使用

# WHAT

見た目の簡略化

# WHY
使いやすく、みやすくするため。

kinshotomoya commented on app/views/groups/edit.html.haml:
```diff
@@ -1,33 +1,40 @@
 .chat-group-form
   %h1 チャットグループ編集
-  %form#edit_chat_group_22.edit_chat_group{"accept-charset" => "UTF-8", :action => "/chat_groups/22", :method => "post"}
+  %form#edit_chat_group_22.edit_chat_group{"accept-charset" => "UTF-8", action: "/chat_groups/22", method: "post" }
```
"accept-charset" => "UTF-8"　ここもシンボルに変えましょ！

kinshotomoya commented on app/views/groups/index.html.haml:
```diff
@@ -6,7 +6,7 @@
         .fa.fa-pencil-square-o{"aria-hidden" => "true"}
       .left_content__chat_groups
-        - current_user.groups.each do |group|
+        - @groups.each do |group|
```
render　collection 使ってください！

kinshotomoya reviewed (commented):
コメントしました！
これは、追加機能と削除機能は、まだ実装されていませんか？？？

下記、修正してください！

avaklee committed:
Incremental search - 追加削除機能

 # WHAT
Incremental search の追加削除と前回の指摘された見た目の部分の修正

# WHY
必要な機能なので、見た目は読みやすく扱いやすくするため

avaklee commented:
追加機能と削除機能は、してなかったのですが、今回のプルリクエストでは追加しました。

avaklee commented:
<img width="1440" alt="2017-06-17 11 00 01" src="https://user-images.githubusercontent.com/29111223/27249261-275648b6-534c-11e7-8e03-a02198346813.png">

seoky0 commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,67 @@
+
+  $('.chat-group-form__search').on('click', '.chat-group-user__btn--add', function() {
+    var user = $(this);
+    var html2 = buildHTML2(user);
```
こちらはhtmlという変数名で大丈夫です。

seoky0 commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,67 @@
+  };
+
+  $('.chat-group-form__search').on('click', '.chat-group-user__btn--add', function() {
+    var user = $(this);
```
thisは追加ボタンを指すので、変数名をつけるならaddBtnなどでしょうか。

seoky0 commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,67 @@
+  })
+
+  function buildHTML2(user){
+    var html2 =
```
変数名はhtmlで大丈夫です

seoky0 commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,67 @@
+    var user = $(this);
+    var html2 = buildHTML2(user);
+    $('.chat-group-form__search.clearfix').append(html2);
+    $(this).parent().remove();
```
せっかく上で変数化しているので変数を使いましょう

seoky0 commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,67 @@
+   });
+
+  $('.chat-group-form__search').on('click', '.js-remove-btn', function() {
+    console.log("test");
```
デバッグ用の記述はレビューに出す際は必ず消しましょう

oyaoyaoya commented on app/assets/javascripts/group.js:
```diff
@@ -32,5 +36,32 @@ $(function(){
+   });
+
+  $('.chat-group-form__search').on('click', '.js-remove-btn', function() {
+    console.log("test");
```
PRでは、console.logは消しましょう！

seoky0 commented on app/controllers/users_controller.rb:
```diff
@@ -1,5 +1,13 @@
+  def index
+    @users = User.chosen(params[:keyword])
+    respond_to do |format|
+      format.html
```
htmlで処理することはないので、消しましょう

oyaoyaoyana commented on app/assets/javascripts/group.js:
```diff
@@ -32,5 +36,32 @@ $(function(){
       $("#textStatus").html("textStatus : " + textStatus);
```
失敗時は代わりに
alert('ユーザーの検索に失敗しました');
等をつけてあげると良いと思います。

oyaoyaoyana commented on app/views/groups/_groups.html.haml:
```diff
@@ -0,0 +1,5 @@
+- @groups.each do |group|
```
もう一度、render collectionに関して調べてみましょう。
更に簡単に書けるはずです。

avaklee replied:
パーシャルの部分ではなく、index html の方を<%= render @groups %>とのことでしょうか？

seoky0 replied:
部分テンプレートの中でeach文を使って繰り返すのではなく、renderのcollectionオプションを使って記述しましょうということですね。
collectionを使うことで、each文を使わずともインスタンス変数の要素の数だけビューを描画してくれます。

oyaoyaoyana commented on app/assets/javascripts/group.js:
```diff
@@ -2,15 +2,19 @@ $(function(){
+    + '<a class="user-search-add chat-group-user__btn chat-group-user__btn--add" data-user-id='
+        + user.id
+    + ' ' +  "data-user-name="
+      + user.name
```
data-user-name=
 user.name
このプロパティはなぜ必要なのでしょうか？

avaklee replied:
上でとっているので、二度取る必要なないとのことを指しているのでしょうか？

avaklee replied:
追加ボタンと押された時にその選ばれたユーザーの名前とidの情報を取得し（ボタンにデーターとして乗せて）、新たに削除ボタンのあるhtmlを作るためにここでuser.nameで取得してます。

seoky0 replied:
良いですね！

oyaoyaoyana reviewed (commented):
追加と削除機能に関しては綺麗に実装ができています！
もう少し考えてみましょう！

avaklee committed:
修正

#WHAT
指摘された部分の修正、また質問
# WHY
見た目や変数の中身をわかりやすくするためのネーミングなど

seoky0 commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,67 @@
+    })
+  })
+
+  function buildHTML2(user){
```
変数も関数も数字をつけてバージョン管理っぽくするとそれが何をする変数や関数なのかわかりづらくなるので、より具体的な命名をすると更に良いですね。
今回はユーザーを追加するHTMLを作るので、buildAddUserHTMLとかですかね

avaklee committed:
修正

# WHAT
Render - collection used

# WHY
Each文で回すより早く効率的

kinshotomoya reviewed (commented):
LGTMです！
マージして次行きましょう！

お疲れ様です！！！！！
